File: tests/diagnose_voice_pipeline.py
import asyncio
import websockets
import json
import logging
import math
import struct
import time

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger("VoiceDiagnose")

async def generate_sine_wave(duration_sec=3, frequency=440, sample_rate=16000):
    """Generates a raw PCM 16-bit mono audio buffer (Sine Wave at 440Hz)."""
    num_samples = int(duration_sec * sample_rate)
    audio_data = bytearray()
    
    for i in range(num_samples):
        # Generate sine wave sample
        sample = 32767 * math.sin(2 * math.pi * frequency * i / sample_rate)
        # Pack as little-endian 16-bit signed integer
        audio_data.extend(struct.pack('<h', int(sample)))
        
    return bytes(audio_data)

async def test_backend_direct():
    """Test backend direct."""
    uri = "ws://127.0.0.1:8005"
    logger.info(f"Connecting to Voice Server at {uri}...")
    
    try:
        async with websockets.connect(uri) as websocket:
            logger.info("✅ Connected to WebSocket.")
            
            # 1. Wait for Ready Signal (optional, based on my server code)
            try:
                msg = await asyncio.wait_for(websocket.recv(), timeout=2.0)
                logger.info(f"Server Greeting: {msg}")
            except asyncio.TimeoutError:
                logger.warning("No greeting received (might be okay depending on protocol).")

            # 2. Generate Synthetic Audio
            logger.info("Generating 3 seconds of synthetic audio (440Hz Sine)...")
            # NOTE: Sine waves are not speech, so transcription might be garbage or empty,
            # BUT the server should at least acknowledge receiving it or log something.
            # Ideally, we'd use a real speech file, but let's test *throughput* first.
            chunk_size = 4096 
            audio_bytes = await generate_sine_wave()
            
            total_chunks = len(audio_bytes) // chunk_size
            logger.info(f"Streaming {len(audio_bytes)} bytes in {total_chunks} chunks...")

            # 3. Stream
            for i in range(0, len(audio_bytes), chunk_size):
                chunk = audio_bytes[i:i+chunk_size]
                await websocket.send(chunk)
                await asyncio.sleep(0.01) # Simulate real-time streaming
                
                # Check for partial responses immediately?
                # Usually client sends all, server responds async.
            
            logger.info("Stream complete. Sending STOP command...")
            await websocket.send(json.dumps({"command": "stop"}))
            
            # 4. Wait for Final Response
            start_wait = time.time()
            while time.time() - start_wait < 10:
                try:
                    response = await websocket.recv()
                    data = json.loads(response)
                    logger.info(f"Received: {data}")
                    
                    if data.get("is_final") or data.get("type") == "transcription":
                        logger.info("✅ Backend responded with transcription packet!")
                        return True
                        
                except websockets.exceptions.ConnectionClosed:
                    logger.error("Connection closed unexpectedly.")
                    break
                except Exception as e:
                    logger.error(f"Error receiving: {e}")
                    break
            
            logger.error("❌ Timed out waiting for transcription response.")
            return False

    except Exception as e:
        logger.error(f"❌ Connection Failed: {e}")
        return False

if __name__ == "__main__":
    asyncio.run(test_backend_direct())

File: rust_output/src/diagnose_voice_pipeline.rs
use anyhow::{Result, Context};
use std::collections::HashMap;
use tokio;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Generates a raw PCM 16-bit mono audio buffer (Sine Wave at 440Hz).
pub async fn generate_sine_wave(duration_sec: String, frequency: String, sample_rate: String) -> () {
    // Generates a raw PCM 16-bit mono audio buffer (Sine Wave at 440Hz).
    let mut num_samples = (duration_sec * sample_rate).to_string().parse::<i64>().unwrap_or(0);
    let mut audio_data = bytearray();
    for i in 0..num_samples.iter() {
        let mut sample = (32767 * math::sin(((((2 * math::pi) * frequency) * i) / sample_rate)));
        audio_data.extend(r#struct::pack("<h".to_string(), sample.to_string().parse::<i64>().unwrap_or(0)));
    }
    bytes(audio_data)
}

/// Test backend direct.
pub async fn test_backend_direct() -> Result<()> {
    // Test backend direct.
    let mut uri = "ws://127.0.0.1:8005".to_string();
    logger.info(format!("Connecting to Voice Server at {}...", uri));
    // try:
    {
        let mut websocket = websockets.connect(uri);
        {
            logger.info("✅ Connected to WebSocket.".to_string());
            // try:
            {
                let mut msg = asyncio.wait_for(websocket.recv(), /* timeout= */ 2.0_f64).await;
                logger.info(format!("Server Greeting: {}", msg));
            }
            // except asyncio.TimeoutError as _e:
            logger.info("Generating 3 seconds of synthetic audio (440Hz Sine)...".to_string());
            let mut chunk_size = 4096;
            let mut audio_bytes = generate_sine_wave().await;
            let mut total_chunks = (audio_bytes.len() / chunk_size);
            logger.info(format!("Streaming {} bytes in {} chunks...", audio_bytes.len(), total_chunks));
            for i in (0..audio_bytes.len()).step_by(chunk_size as usize).iter() {
                let mut chunk = audio_bytes[i..(i + chunk_size)];
                websocket.send(chunk).await;
                asyncio.sleep(0.01_f64).await;
            }
            logger.info("Stream complete. Sending STOP command...".to_string());
            websocket.send(serde_json::to_string(&HashMap::from([("command".to_string(), "stop".to_string())])).unwrap()).await;
            let mut start_wait = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
            while (std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64() - start_wait) < 10 {
                // try:
                {
                    let mut response = websocket.recv().await;
                    let mut data = serde_json::from_str(&response).unwrap();
                    logger.info(format!("Received: {}", data));
                    if (data.get(&"is_final".to_string()).cloned() || data.get(&"type".to_string()).cloned() == "transcription".to_string()) {
                        logger.info("✅ Backend responded with transcription packet!".to_string());
                        true
                    }
                }
                // except websockets.exceptions::ConnectionClosed as _e:
                // except Exception as e:
            }
            logger.error("❌ Timed out waiting for transcription response.".to_string());
            false
        }
    }
    // except Exception as e:
}
